controllers/ciresource_fsm.go

```diff
@@ -220,9 +222,12 @@ func (f *CIResourceFSM) handleStateDelete(context CIResourceFSMContext) (time.Du
 	f.logger.Info("removing resource", "Id", context.CIResource.Status.ResourceId)
 
 	if controllerutil.ContainsFinalizer(context.CIResource, ofcirv1.CIResourceFinalizer) {
-		if err := context.Provider.Release(context.CIResource.Status.ResourceId); err != nil {
-			if !errors.As(err, &providers.ResourceNotFoundError{}) {
-				return defaultCIPoolRetryDelay, err
+		// only call the provider if the resource is not a fallback dummy
+		if !context.CIPool.IsFallbackPool() && context.CIResource.Status.ResourceId != fallbackResourceID {
```

Comment:
In order to have a better happy path (and reduce the indentation), it'd be recommended have this check and eventually return immediately. I guess that this will leave the resource in delete status, which is not correct. A proper way to handle such cases would be to implement a validating webhook, but it's currently out of scope.
A reasonable compromise in such case could be to log the erroneous request (a delete for a not-yet-assigned fallback resource) and move it back to the available

Reply:
I don't understand the part about "move it back to the available". If I want to reduce the pool size, I must delete the resource.

Reply:
Sorry my fault, you're perfectly right - I was thinking to another case (allowed states transitions) which effectively doesn't make sense for such scenario. Fully agree that the resource must be deleted.

For what regards the happy path, fr the sake of clarity the suggested change could be something like the following: 
```
...
if controllerutil.ContainsFinalizer(context.CIResource, ofcirv1.CIResourceFinalizer) {
		
	if context.IsFallbackPool() && context.CIResource.Status.ResourceId != fallbackResourceID {
		controllerutil.RemoveFinalizer(context.CIResource, ofcirv1.CIResourceFinalizer)
		return f.UpdateResourceOnly()
	}

	if err := context.Provider.Release(context.CIResource.Status.ResourceId); err != nil {
		if !errors.As(err, &providers.ResourceNotFoundError{}) {
			return defaultCIPoolRetryDelay, err
		}
	}

	controllerutil.RemoveFinalizer(context.CIResource, ofcirv1.CIResourceFinalizer)
	return f.UpdateResourceOnly()
}
```

Reply:
fixed